oak_functions/client/rust/src/attestation.rs

```diff
@@ -81,45 +81,50 @@ impl Client {
 
 /// gRPC Attestation Service client implementation.
 pub struct AttestationClient {
-    client: Client,
+    channel: GrpcChannel,
     encryptor: AeadEncryptor,
 }
 
 impl AttestationClient {
     pub async fn create(uri: &str, expected_tee_measurement: &[u8]) -> anyhow::Result<Self> {
-        let mut client = Client::create(uri)
+        let mut channel = GrpcChannel::create(uri)
             .await
             .context("Couldn't create gRPC client")?;
-        let encryptor = Self::attest(&mut client, expected_tee_measurement)
+        let encryptor = Self::attest_server(&mut channel, expected_tee_measurement)
             .await
             .context("Couldn't attest server")?;
 
-        Ok(Self { client, encryptor })
+        Ok(Self { channel, encryptor })
     }
 
     /// Attests server for a single gRPC streaming request.
-    async fn attest(
-        client: &mut Client,
+    async fn attest_server(
+        channel: &mut GrpcChannel,
         expected_tee_measurement: &[u8],
     ) -> anyhow::Result<AeadEncryptor> {
-        let attestation_engine =
-            AttestationEngine::<UnattestedPeer>::create(expected_tee_measurement)
-                .context("Couldn't create attestation state machine")?;
-        let identity = attestation_engine
-            .identity()
-            .context("Couldn't get server identity")?;
-
-        // Create attestation request with client's public key.
+        let attestation_engine = AttestationEngine::<attestation::Client, Initializing>::new(
+            AttestationBehavior::create_peer_attestation(expected_tee_measurement)
+                .context("Couldn't create peer attestation behavior")?,
+        );
+
+        // Send attestation initialization message.
+        let (attestation_init, attestation_engine) = attestation_engine
```

Comment:
I am not sure I understand the point of AttestationEngine always returning another instance of AttestationEngine. What is the advantage of doing it this way?

Reply:
Since each state is a new class, and transition functions take `self`, we need to create this class and return it.

Reply:
The advantage is that we don't have an internal variable with a state. And all the errors with incorrect state transitions will be caught by a compiler (and not by `if state == Attested`).